### PR TITLE
change super_schema to use a CommitItr

### PR DESCRIPTION
**Pull request by bheni**

### CODE REVIEW

File: go/libraries/doltcore/doltdb/commit_itr.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	"context"
+	"errors"
+	"github.com/liquidata-inc/dolt/go/store/hash"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+// CommitItr is an interface for iterating over a set of unique commits
+type CommitItr interface {
+	// Next returns the hash of the next commit, and a pointer to that commit.  Implementations of Next must handle
+	// making sure the list of commits returned are unique.  When complete Next will return hash.Hash{}, nil, nil
```
Comment (zachmu) [inline go/libraries/doltcore/doltdb/commit_itr.go]: Returning io.EOF is more idiomatic for iterators in go, right?

File: go/libraries/doltcore/doltdb/commit_itr.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	ddb         *DoltDB
+	rootCommits []*Commit
+	currentRoot int
+
+	added       map[hash.Hash]bool
+	unprocessed []hash.Hash
+	curr        *Commit
+}
+
+// CommitItrForAllBranches returns a CommitItr which will iterate over all commits in all branches in a DoltDB
```
Comment (zachmu) [inline go/libraries/doltcore/doltdb/commit_itr.go]: Needs more detail about the order that results are returned in. Looks like this is a depth-first ancestor traversal for each top-level ref?

File: go/libraries/doltcore/doltdb/commit_itr.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		}
+
+		rootCommits = append(rootCommits, cm)
+	}
+
+	cmItr := CommitItrForRoots(ddb, rootCommits...)
+	return cmItr, nil
+}
+
+// CommitItrForRoots will return a
```
Comment (zachmu) [inline go/libraries/doltcore/doltdb/commit_itr.go]: Suspense is killing me

File: go/libraries/doltcore/doltdb/commit_itr.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+func CommitItrForRoots(ddb *DoltDB, rootCommits ...*Commit) CommitItr {
+	return &commitItr{
+		ddb:         ddb,
+		rootCommits: rootCommits,
+		added:       make(map[hash.Hash]bool, 4096),
+		unprocessed: make([]hash.Hash, 0, 4096),
+	}
+}
+
+// Next returns the hash of the next commit, and a pointer to that commit.  It handles making sure the list of commits
```
Comment (zachmu) [inline go/libraries/doltcore/doltdb/commit_itr.go]: Probably can omit the comment here, or just refer to the interface method.

File: go/libraries/doltcore/doltdb/commit_itr.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+// returned are unique.  When complete Next will return hash.Hash{}, nil, nil
+func (cmItr *commitItr) Next(ctx context.Context) (hash.Hash, *Commit, error) {
+	for cmItr.curr == nil {
+		if cmItr.currentRoot >= len(cmItr.rootCommits) {
+			return hash.Hash{}, nil, nil
```
Comment (zachmu) [inline go/libraries/doltcore/doltdb/commit_itr.go]: Same comment about io.EOF

File: go/libraries/doltcore/doltdb/commit_itr.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+	numUnprocessed := len(cmItr.unprocessed)
+
+	if numUnprocessed == 0 {
+		cmItr.curr = nil
+		cmItr.currentRoot++
+		return cmItr.Next(ctx)
+	}
+
+	next := cmItr.unprocessed[numUnprocessed-1]
```
Comment (zachmu) [inline go/libraries/doltcore/doltdb/commit_itr.go]: Any reason to pop off the end of the list like this, instead of FIFO?

Comment (bheni) [reply go/libraries/doltcore/doltdb/commit_itr.go]: 1. performance.
2. maintain depth first ordering.

File: go/libraries/doltcore/doltdb/commit_itr.go (hunk 1 above)
Comment (bheni) [reply go/libraries/doltcore/doltdb/commit_itr.go]: done

File: go/libraries/doltcore/doltdb/commit_itr.go (hunk 2 above)
Comment (bheni) [reply go/libraries/doltcore/doltdb/commit_itr.go]: I don't think I want to guarantee an traversal order.  Users should just know they are going to get all unique commits.

File: go/libraries/doltcore/doltdb/commit_itr.go (hunk 3 above)
Comment (bheni) [reply go/libraries/doltcore/doltdb/commit_itr.go]: updated

File: go/libraries/doltcore/doltdb/commit_itr.go (hunk 4 above)
Comment (bheni) [reply go/libraries/doltcore/doltdb/commit_itr.go]: I think the generated go docs prefer this be here.

File: go/libraries/doltcore/doltdb/commit_itr.go (hunk 5 above)
Comment (bheni) [reply go/libraries/doltcore/doltdb/commit_itr.go]: updated

**Pull request merged**